Firestore/core/src/firebase/firestore/model/transform_operations.h

```diff
@@ -233,6 +245,116 @@ class ArrayTransform : public TransformOperation {
   }
 };
 
+/**
+ * Implements the backend semantics for locally computed NUMERIC_ADD (increment)
+ * transforms. Converts all field values to longs or doubles and resolves
+ * overflows to LONG_MAX/LONG_MIN.
+ */
+class NumericIncrementTransform : public TransformOperation {
+ public:
+  NumericIncrementTransform(FSTNumberValue* operand)
+      : operand_(std::move(operand)) {
+  }
+
+  ~NumericIncrementTransform() override {
+  }
+
+  Type type() const override {
+    return Type::Increment;
+  }
+
+  FSTFieldValue* ApplyToLocalView(
+      FSTFieldValue* previousValue,
+      FIRTimestamp* /* localWriteTime */) const override {
+    // Return an integer value only if the previous value and the operand is an
+    // integer.
+    if ([previousValue isKindOfClass:[FSTIntegerValue class]] &&
+        [operand_ isKindOfClass:[FSTIntegerValue class]]) {
+      int64_t sum = SafeIncrement(
+          (reinterpret_cast<FSTIntegerValue*>(previousValue)).internalValue,
+          (reinterpret_cast<FSTIntegerValue*>(operand_)).internalValue);
+      return [FSTIntegerValue integerValue:sum];
+    } else if ([previousValue isKindOfClass:[FSTIntegerValue class]]) {
+      double sum =
+          (reinterpret_cast<FSTIntegerValue*>(previousValue)).internalValue +
+          OperandAsDouble();
+      return [FSTDoubleValue doubleValue:sum];
+    } else if ([previousValue isKindOfClass:[FSTDoubleValue class]]) {
+      double sum =
+          (reinterpret_cast<FSTDoubleValue*>(previousValue)).internalValue +
+          OperandAsDouble();
+      return [FSTDoubleValue doubleValue:sum];
+    } else {
+      // If the existing value is not a number, use the value of the transform
+      // as the new base value.
+      return operand_;
+    }
+  }
+
+  FSTFieldValue* ApplyToRemoteDocument(
+      FSTFieldValue* previousValue,
+      FSTFieldValue* transformResult) const override {
+    return transformResult;
+  }
+
+  FSTNumberValue* operand() const {
+    return operand_;
+  }
+
+  bool idempotent() const override {
+    return false;
+  }
+
+  bool operator==(const TransformOperation& other) const override {
+    if (other.type() != type()) {
+      return false;
+    }
+    auto numeric_add = static_cast<const NumericIncrementTransform&>(other);
+    return [operand_ isEqual:numeric_add.operand_];
+  }
+
+  // For Objective-C++ hash; to be removed after migration.
+  // Do NOT use in C++ code.
+  NSUInteger Hash() const override {
+    NSUInteger result = 37;
+    result = 31 * result + [operand_ hash];
+    return result;
+  }
+
+ private:
+  FSTNumberValue* operand_;
+
+  /**
+   * Implements integer addition. Overflows are resolved to LONG_MAX/LONG_MIN.
+   */
+  int64_t SafeIncrement(int64_t x, int64_t y) const {
+    int64_t r = x + y;
```

Comment:
This is undefined behavior, and it's not safe to rely on any particular value of `r`. See [this thread](https://stackoverflow.com/questions/199333/how-to-detect-integer-overflow) for some suggestions. Alternatively, a larger int type would be helpful, but unfortunately Abseil only has [`uint128`](https://github.com/abseil/abseil-cpp/blob/master/absl/numeric/int128.h), not `int128`, for the time being (I guess a workaround would be to use modulo, but it might end up harder to implement than checking).